src/txn/txn_rollback_to_stable.c

```diff
@@ -0,0 +1,403 @@
+
+	/*
+	 * Immediately durable files don't get their commits wiped. This case
+	 * mostly exists to support the semantic required for the oplog in
+	 * MongoDB - updates that have been made to the oplog should not be
+	 * aborted. It also wouldn't be safe to roll back updates for any
+	 * table that had it's records logged, since those updates would be
+	 * recovered after a crash making them inconsistent.
+	 */
+	if (__wt_btree_immediately_durable(session)) {
+		/*
+		 * Add the btree ID to the bitstring, so we can exclude any
+		 * lookaside entries for this btree.
+		 */
+		__bit_set(
+		    S2C(session)->stable_rollback_bitstring, btree->id);
+		return (0);
+	}
+
+	/* There is never anything to do for checkpoint handles */
+	if (session->dhandle->checkpoint != NULL)
+		return (0);
+
+	/* There is nothing to do on an empty tree. */
+	if (btree->root.page == NULL)
+		return (0);
+
+	if (btree->type != BTREE_ROW)
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable "
+		    "is only supported for row store btrees");
+
+	/*
+	 * Copy the stable timestamp, otherwise we'd need to lock it each time
+	 * it's accessed. Even though the stable timestamp isn't supposed to be
+	 * updated while rolling back, accessing it without a lock would
+	 * violate protocol.
+	 */
+	__wt_readlock(session, &txn_global->rwlock);
+	__wt_timestamp_set(&rollback_timestamp, &txn_global->stable_timestamp);
+	__wt_readunlock(session, &txn_global->rwlock);
+
+	/*
+	 * Ensure the eviction server is out of the file - we don't
+	 * want it messing with us. This step shouldn't be required, but
+	 * it simplifies some of the reasoning about what state trees can
+	 * be in.
+	 */
+	WT_RET(__wt_evict_file_exclusive_on(session));
+	ret = __txn_rollback_to_stable_btree_walk(
+	    session, &rollback_timestamp);
+	__wt_evict_file_exclusive_off(session);
+
+	return (ret);
+}
+
+/*
+ * __txn_rollback_to_stable_check --
+ *	Ensure the rollback request is reasonable.
+ */
+static int
+__txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
+{
+	WT_TXN_GLOBAL *txn_global;
+	WT_TXN_STATE *s;
+	uint32_t i, session_cnt;
+	bool stable_set;
+
+	txn_global = &S2C(session)->txn_global;
+	__wt_readlock(session, &txn_global->rwlock);
+	stable_set = !__wt_timestamp_iszero(&txn_global->stable_timestamp);
+	__wt_readunlock(session, &txn_global->rwlock);
+	if (!stable_set)
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable requires a "
+		    "stable timestamp");
+
+	/*
+	 * Help the user - see if they have any active transactions. I'd
+	 * like to check the transaction running flag, but that would
+	 * require peeking into all open sessions, which isn't really
+	 * kosher.
+	 */
+	WT_ORDERED_READ(session_cnt, S2C(session)->session_cnt);
+	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++) {
+		if (s->id != WT_TXN_NONE || s->pinned_id != WT_TXN_NONE)
+			WT_RET_MSG(session, EINVAL,
+			    "rollback_to_stable not supported with "
+			    "active transactions");
+	}
```

Comment:
I think this check is very similar to what is required for WT-3417.  We should share code.

FTR, I think this should be equivalent to `__wt_txn_update_oldest` and checking that `txn_global->oldest_id == txn_global->current` but will check that in WT-3417.